File: src/contests/cg/puzzle/pascal/sum-of-divisors.pas
// 🎮 CodinGame Puzzle - sum-of-divisors
// https://www.codingame.com/training/medium/sum-of-divisors

{$mode objfpc}{$H+}
var
  n, total, d: int64;
begin
  // d appears as a divisor in floor(n/d) of the numbers 1..n, so the total
  // sum of divisors is sum over d of d * floor(n/d).
  readln(n);
  total := 0;
  d := 1;
  while d <= n do
  begin
    total := total + d * (n div d);
    d := d + 1;
  end;
  writeln(total);
end.

File: src/contests/cg/puzzle/haskell/sum-of-divisors.hs
-- 🎮 CodinGame Puzzle - sum-of-divisors
-- https://www.codingame.com/training/medium/sum-of-divisors

-- d appears as a divisor in floor(n/d) of the numbers 1..n, so the total sum
-- of divisors is sum over d of d * floor(n/d).
main :: IO ()
main = do
  n <- readLn :: IO Integer
  print (sum [d * (n `div` d) | d <- [1 .. n]])
